### PR TITLE
validate blocklist URL before adding to the database

### PR DESCRIPTION
**Pull request by PromoFaux**
**By submitting this pull request, I confirm the following:** 

- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md), as well as this entire template.
- [x] I have made only one major change in my proposed changes.
- [ ] I have commented my proposed changes within the code.
- [x] I have tested my proposed changes, and have included unit tests where possible.
- [x] I am willing to help maintain this change if there are issues with it later.
- [x] I give this submission freely and claim no ownership.
- [x] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [x] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

---
**What does this PR aim to accomplish?:**
Prevents an invalid URL from being added to the adlist table

### CODE REVIEW

File: advanced/Scripts/webpage.sh [hunk 1]
```diff
@@ -401,22 +401,37 @@ SetWebUILayout() {
...
+            sqlite3 "${gravityDBfile}" "INSERT OR IGNORE INTO adlist (address, comment) VALUES ('${address}', '${comment}')"
+        elif [[ "${args[2]}" == "del" ]]; then
+            sqlite3 "${gravityDBfile}" "DELETE FROM adlist WHERE address = '${address}'"
+        else
+            echo "Not permitted"
+            return 1
+        fi
     else
-        echo "Not permitted"
+        echo "Invald Url"
```
Comment (DL6ER) [inline advanced/Scripts/webpage.sh]: Typo

File: advanced/Scripts/webpage.sh [hunk 2]
```diff
@@ -401,22 +401,37 @@ SetWebUILayout() {
     change_setting "WEBUIBOXEDLAYOUT" "${args[2]}"
 }
 
+CheckUrl(){
+    local url validUrl regex
+    regex="^(?:http(s)?:\/\/)?[\w.-]+(?:\.[\w\.-]+)+[\w\-\._~:/?#[\]@!\$&'\(\)\*\+,;=.]+$"
```
Comment (DL6ER) [inline advanced/Scripts/webpage.sh]: Did we use this regex before or is it a new one?

Comment (PromoFaux) [reply advanced/Scripts/webpage.sh]: It's a stack-exchange special, of course. Have we ever validated URLs in bash before? I could not see anything.

Comment (DL6ER) [reply advanced/Scripts/webpage.sh]: I don't know, could you add the SE URL there so we can find the source more easily when needed in the future?

File: advanced/Scripts/webpage.sh [hunk 3]
```diff
@@ -401,22 +401,38 @@ SetWebUILayout() {
     change_setting "WEBUIBOXEDLAYOUT" "${args[2]}"
 }
 
+CheckUrl(){
+    local url validUrl regex
+    #regex from https://www.regextester.com/94502
+    regex="^(?:http(s)?:\/\/)?[\w.-]+(?:\.[\w\.-]+)+[\w\-\._~:/?#[\]@!\$&'\(\)\*\+,;=.]+$"
```
Comment (DL6ER) [inline advanced/Scripts/webpage.sh]: We also support `file://` as protocol, maybe even `ftp://` (?). This regex only permits `http(s)://` or nothing.

Comment (dschaper) [reply advanced/Scripts/webpage.sh]: Possibly `///` shorthand as well.

**Comment by PromoFaux** [comment]
I thought I had mentioned in the original PR comment/commits..

But this PR was to a fix for CVE-2020-11108 discovered by @frichetten

You can see his post about it here:

https://frichetten.com/blog/cve-2020-11108-pihole-rce/

**Pull request merged**